### PR TITLE
Add check verifying labels and annotation are not confused (#5718)

### LINKED ISSUE #5718: Check should warn when expected labels and annotations are confused
I just lost a lot of time to setting the `mirror.linkerd.io/exported=true` annotation, when it should have been a label.

It would be _awesome_ if `linkerd check` could flag this sort of thing. Off the top of my head:

* `mirror.linkerd.io/exported` should never be an annotation
* `config.linkerd.io/*` should never be a label
* `linkerd.io/inject` should never be a label

There may be others?

### PR DESCRIPTION
**Pull request by Szymongib**
No indication that labels and annotations were set incorrectly.

Check if labels set by user are not set as annotations and vice versa.

Unit tests and manual test.

Fixes #5718

Signed-off-by: Szymon Gibała <szymongib@gmail.com>

### CODE REVIEW

**Review by alpeb — COMMENTED** [review]
Looking good :+1: 
What do you think about moving the functions in `healthcheck.go` to a new file like `healthcheck_labels.go` so we keep this file size under control?
Also it'd be great to add a test case into `healthcheck_test.go` with a couple of resources violating these checks so we can ensure the output of the checks is right :wink:

File: pkg/healthcheck/healthcheck.go [hunk 1]
```diff
@@ -1342,6 +1342,32 @@ func (hc *HealthChecker) allCategories() []Category {
 						return nil
 					},
 				},
+				{
+					description: "data plane pods labels check",
```
Comment (dadjeibaah) [inline pkg/healthcheck/healthcheck.go:1346]: I think we can change this description's wording to match the other checks. How about:
`data plane pod labels are configured correctly`

File: pkg/healthcheck/healthcheck.go [hunk 2]
```diff
@@ -1342,6 +1342,32 @@ func (hc *HealthChecker) allCategories() []Category {
...
+						pods, err := hc.GetDataPlanePods(ctx)
+						if err != nil {
+							return err
+						}
+
+						return checkMisconfiguredPodsLabels(pods)
+					},
+				},
+				{
+					description: "data plane services labels and annotations check",
```
Comment (dadjeibaah) [inline pkg/healthcheck/healthcheck.go:1359]: Same goes for this description as well. How about:
`data plane service labels and annotations are configured correctly`

File: pkg/healthcheck/healthcheck_test.go [hunk 3]
```diff
@@ -1898,6 +1898,145 @@ func TestValidateDataPlanePods(t *testing.T) {
 		}
 	})
 }
+
+func TestDataPlanePodLabels(t *testing.T) {
```
Comment (dadjeibaah) [inline pkg/healthcheck/healthcheck_test.go:1902]: Thanks for adding tests 🙏 !!

File: pkg/healthcheck/healthcheck.go [hunk 4]
```diff
@@ -2522,3 +2557,101 @@ func containsProxy(pod corev1.Pod) bool {
...
+	for _, pod := range pods {
+		invalidLabels := getMisconfiguredLabels(pod.ObjectMeta)
+		if len(invalidLabels) > 0 {
+			invalid = append(invalid,
+				fmt.Sprintf("\t* pod %s/%s contains labels, which should be annotations: %s", pod.Namespace, pod.Name, strings.Join(invalidLabels, ", ")))
+		}
+	}
+
+	if len(invalid) > 0 {
+		return fmt.Errorf("Some labels on pods might be missconfigured:\n%s", strings.Join(invalid, "\n"))
```
Comment (dadjeibaah) [inline pkg/healthcheck/healthcheck.go:2586]: I was playing around with the messaging around here to be more informative and let users know what exactly is misconfigured. I think this formats the error message in a way that a user can act upon it.

```suggestion
			invalid = append(invalid,
				fmt.Sprintf("\t* %s/%s\n\t\t%s", pod.Namespace, pod.Name, strings.Join(invalidLabels, "\n\t\t")))
		}
	}

	if len(invalid) > 0 {
		return fmt.Errorf("Some labels on data plane pods should be annotations:\n%s", strings.Join(invalid, "\n"))
```

The output looks like this:
```bash
‼ data plane and cli versions match
    emoji-5f947b5dbd-44fkm running git-68cebe88 but cli running dev-undefined
    see https://linkerd.io/checks/#l5d-data-plane-cli-version for hints
‼ data plane pod labels are configured correctly
    Some labels on data plane pods should be annotations:
        * emojivoto/emoji-5f947b5dbd-44fkm
                linkerd.io/inject
        * emojivoto/voting-f7f875999-phcxv
                linkerd.io/inject
                config.linkerd.io/multicluster-option
                config.linkerd.io/viz-option
    see https://linkerd.io/checks/#l5d-data-plane-pod-labels for hints
√ data plane service labels and annotations are configured correctly
```

**Review by dadjeibaah — COMMENTED** [review]
@Szymongib, thanks for putting this PR together! This is going to be super useful. I think the change is looking great. I left a few suggestions for your consideration.

**Comment by Szymongib** [comment]
Thanks a lot for review @alpeb and @dadjeibaah!

@alpeb a questions regarding your comment
> Also it'd be great to add a test case into `healthcheck_test.go` with a couple of resources violating these checks so we can ensure the output of the checks is right 😉

There are already some tests violating the check that expect the error https://github.com/linkerd/linkerd2/pull/5943/files#diff-932ac70a02fcd5e0e4b89b0a03f1302aeac2d162ea1b7ae7a074839d24e6dbf5R1921 do you suggest to add assertion of the actual error message?

**Comment by alpeb** [comment]
Ah sorry I missed that, but yeah, it'd be nice to check the errors themselves as we do in the other negative tests in that file.

**Comment by Szymongib** [comment]
I have adjusted PR to address your comments, let me know if you have any other concerns or see ways to improve.

Changes:
- Move labels and annotations checks to `healthcheck_labels.go` file
- Split services check to two to align error messages with pods check - one check for labels the other for annotations
- Change error messages according to suggestions
- Check if error message is correct in tests

File: pkg/healthcheck/healthcheck.go [hunk 5]
```diff
@@ -1353,6 +1353,45 @@ func (hc *HealthChecker) allCategories() []Category {
...
+						pods, err := hc.GetDataPlanePods(ctx)
+						if err != nil {
+							return err
+						}
+
+						return checkMisconfiguredPodsLabels(pods)
+					},
+				},
+				{
+					description: "data plane service labels configured correctly",
```
Comment (dadjeibaah) [inline pkg/healthcheck/healthcheck.go:1370]:
```suggestion
					description: "data plane service labels are configured correctly",
```

**Review by dadjeibaah — COMMENTED** [review]
This looks great! Thanks for making those updates. I had one minor nit but that's pretty much it

File: pkg/healthcheck/healthcheck.go (hunk 5 above)
Comment (Szymongib) [reply pkg/healthcheck/healthcheck.go]: Ah, thanks! I missed that!

**Review by alpeb — APPROVED** [review]
Thanks @Szymongib , looks great! :+1:

**Review by Pothulapati — APPROVED** [review]
LGTM! Thanks for taking this up! 🔥 @Szymongib

Can you also update the `linkerd/website` repo with the troubleshooting docs for the newly added hint anchors? They could just mention an explanation on what labels and annotations the CLI expects for pods and services?

**Comment by Szymongib** [comment]
Hi @Pothulapati, sure!
I have created the PR https://github.com/linkerd/website/pull/1046 let me know if that is correct place and if description is sufficient.

**Pull request merged**